### PR TITLE
Dask imperative plays well with other collections.

### LINKED ISSUE #497: Profiler fails with KeyError with dask.bag + dask.do
Here's a snippet that reproduces the issue

```python
import dask.bag as db
from dask import do
from dask.diagnostics import Profiler


def inc(x):
    return x + 1


@do
def dec(seq):
    return [x - 1 for x in seq]

bag = dec(db.from_sequence([1, 2, 3]).map(inc))

result = bag.compute()
assert result == [1, 2, 3]  # works

with Profiler() as prof:
    result = bag.compute()  # somehow fails because of the profiler
```

I suspect this is related to using both thread and process based schedulers

### PR DESCRIPTION
**Pull request by jcrist**
```python
>>> import dask.array as da
>>> import dask.bag as db
>>> from dask.imperative import do
>>> from dask.diagnostics import ProgressBar
>>> import numpy as np

# Create a list of 2 numpy arrays, 2 dask arrays, and a bag
>>> arr1 = np.arange(100).reshape((10, 10))
>>> arr2 = arr1.dot(arr1.T)
>>> darr1 = da.from_array(arr1, chunks=(5, 5))
>>> darr2 = da.from_array(arr2, chunks=(5, 5))
>>> b = db.from_sequence([1, 2, 3])
>>> seq = [arr1, arr2, darr1, darr2, b]
# Sum the sums of each element in the list
# Inner sum is done by each collection, outer sum is done by imperative
>>> out = do(sum)([i.sum() for i in seq])
# Compute all at once. All dasks from all collections are merged.
>>> with ProgressBar():
...     out.compute()
[########################################] | 100% Completed |  0.1s
```

Fixes #497.

### CODE REVIEW

**Comment by jcrist** [comment]
cc @cpcloud.

File: dask/imperative.py [hunk 1]
```diff
@@ -193,13 +196,11 @@ def compute(*args, **kwargs):
...
     (3, 4)
-    >>> compute(a, [b, c])
+    >>> compute(a, [b, c])  # Works for lists of Values
     (1, [3, 4])
     """
     args = [value(a) for a in args]
-    dsk = merge(*[arg.dask for arg in args])
-    keys = [arg.key for arg in args]
-    return tuple(get(dsk, keys, **kwargs))
+    return base.compute(*args, **kwargs)
```
Comment (jcrist) [inline dask/imperative.py]: One question that came up while implementing this was whether this should be default behavior for `base.compute`. I can see the use of being able to say `compute([(val1, val2), (val3, val4)])`, and get results in the same structure as the input. I'm not sure if this same behavior is useful for the other collections, but a single common `compute` function across all of dask could be nice.

File: dask/imperative.py [hunk 2]
```diff
@@ -63,6 +63,12 @@ def to_task_dasks(expr):
     """
     if isinstance(expr, Value):
         return expr.key, expr._dasks
+    elif isinstance(expr, base.Base):
+        name = tokenize(str(expr), True)
+        keys = expr._keys()
+        dsk = expr._optimize(expr.dask, keys)
+        dsk[name] = (expr._finalize, expr, (list, keys))
```
Comment (mrocklin) [inline dask/imperative.py]: Beware that `keys` may be a list of lists.  There is probably a `concrete` function lying around that might replace `list` here.

**Comment by jcrist** [comment]
I think this is good to go in. If we want to merge the two `compute` functions, I can take care of that later.

File: dask/imperative.py [hunk 3]
```diff
@@ -63,6 +64,12 @@ def to_task_dasks(expr):
     """
     if isinstance(expr, Value):
         return expr.key, expr._dasks
+    elif isinstance(expr, base.Base):
+        name = tokenize(str(expr), True)
+        keys = expr._keys()
+        dsk = expr._optimize(expr.dask, keys)
+        dsk[name] = (expr._finalize, expr, (concrete, keys))
```
Comment (mrocklin) [inline dask/imperative.py]: Whoa, it's a little surreal to have the collection be in the graph itself.

File: dask/imperative.py (hunk 1 above)
Comment (mrocklin) [inline dask/imperative.py]: I can imagine this being convenient.

File: dask/imperative.py (hunk 3 above)
Comment (jcrist) [inline dask/imperative.py]: This is needed because `dask.array` needs to know the shape of the output to properly finalize it.

**Pull request merged**